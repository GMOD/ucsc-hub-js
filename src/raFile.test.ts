// @ts-nocheck
import fs from 'fs'
import RaFile from './raFile'

describe('RaFile reader', () => {
  it('creates an empty ra file', () => {
    const raFile = new RaFile()
    expect(raFile).toMatchInlineSnapshot('Map {}')
    expect(raFile._stanzaAndCommentOrder).toEqual([])
    expect(raFile.nameKey).toBeUndefined()
    expect(raFile.toString()).toEqual('')
  })

  it('populates an empty ra file', () => {
    const raFile = new RaFile()
    const input1 = 'key1 valueA\nkey2 valueB\n'
    const comment = '# A comment\n'
    const input2 = 'key1 valueC\nkey2 valueD\n'
    raFile.add(input1)
    raFile.add(comment)
    raFile.add(input2)
    expect(raFile).toMatchSnapshot()
    expect(raFile._stanzaAndCommentOrder).toEqual([
      'valueA',
      '# A comment',
      'valueC',
    ])
    expect(raFile.nameKey).toEqual('key1')
    expect(raFile.toString()).toEqual(`${input1}\n${comment}\n${input2}`)
  })

  it('parses a simple three stanza file', () => {
    const input = fs.readFileSync('test/basic.ra', 'utf8')
    const raFile = new RaFile(input)
    expect(raFile).toMatchSnapshot()
    expect(raFile._stanzaAndCommentOrder).toEqual(['valA', 'valD', 'valG'])
    expect(raFile.nameKey).toEqual('key1')
    expect(raFile.toString()).toEqual(input)
  })

  it('parses a file with comments', () => {
    const input = fs.readFileSync('test/comments.ra', 'utf8')
    const raFile = new RaFile(input)
    expect(raFile).toMatchSnapshot()
    expect(raFile._stanzaAndCommentOrder).toEqual([
      '#comment1\n#comment2',
      'valA',
      'valD',
    ])
    expect(raFile.nameKey).toEqual('key1')
    expect(raFile.toString()).toEqual(input)
  })

  it('handles CRLF newlines', () => {
    const input =
      'key1 valueA\r\n' +
      'key2 valueB\r\n\r\n' +
      '# A comment\r\n\r\n' +
      'key1 valueC\r\n' +
      'key2 valueD\r\n'
    const raFile = new RaFile(input)
    expect(raFile).toMatchSnapshot()
    expect(raFile._stanzaAndCommentOrder).toEqual([
      'valueA',
      '# A comment',
      'valueC',
    ])
    expect(raFile.nameKey).toEqual('key1')
    expect(raFile.toString()).toEqual(input.replace(/\r/g, ''))
  })

  it('parses a list of stanzas', () => {
    const input = [
      'key1 valueA\nkey2 valueB\n',
      '# A comment\n',
      'key1 valueC\nkey2 valueD\n',
    ]
    const raFile = new RaFile(input)
    expect(raFile).toMatchSnapshot()
    expect(raFile._stanzaAndCommentOrder).toEqual([
      'valueA',
      '# A comment',
      'valueC',
    ])
    expect(raFile.nameKey).toEqual('key1')
    expect(raFile.toString()).toEqual(input.join('\n'))
  })

  it('handles indented stanzas', () => {
    const input = fs.readFileSync('test/indented.ra', 'utf8')
    const raFile = new RaFile(input)
    expect(raFile).toMatchSnapshot()
    expect(raFile._stanzaAndCommentOrder).toEqual(['valA', 'valD', 'valG'])
    expect(raFile.nameKey).toEqual('key1')
    expect(raFile.toString()).toEqual(input)
  })

  it('adds a stanza', () => {
    const raFile = new RaFile(fs.readFileSync('test/basic.ra', 'utf8'))
    raFile.add('key1 valJ\nkey2 valK\nkey3 valL\n')
    expect(raFile).toMatchSnapshot()
    expect(raFile._stanzaAndCommentOrder).toEqual([
      'valA',
      'valD',
      'valG',
      'valJ',
    ])
    expect(raFile.nameKey).toEqual('key1')
    expect(raFile.toString()).toEqual(
      fs.readFileSync('test/afterAdd.ra', 'utf8'),
    )
  })

  it('deletes a stanza', () => {
    const raFile = new RaFile(fs.readFileSync('test/basic.ra', 'utf8'))
    raFile.delete('valD')
    raFile.delete('nonexistent')
    expect(raFile).toMatchSnapshot()
    expect(raFile._stanzaAndCommentOrder).toEqual(['valA', 'valG'])
    expect(raFile.nameKey).toEqual('key1')
    expect(raFile.toString()).toEqual(
      fs.readFileSync('test/afterDelete.ra', 'utf8'),
    )
  })

  it('clears', () => {
    const raFile = new RaFile(fs.readFileSync('test/basic.ra', 'utf8'))
    raFile.clear()
    expect(raFile).toMatchSnapshot()
    expect(raFile._stanzaAndCommentOrder).toEqual([])
    expect(raFile.nameKey).toBeUndefined()
    expect(raFile.toString()).toEqual('')
  })

  it('updates a stanza', () => {
    const input = fs.readFileSync('test/basic.ra', 'utf8')
    const raFile = new RaFile(input)
    const updatedStanza = raFile.get('valD')
    updatedStanza.indent = '    '
    raFile.set('valD', updatedStanza)
    expect(raFile).toMatchSnapshot()
    expect(raFile._stanzaAndCommentOrder).toEqual(['valA', 'valD', 'valG'])
    expect(raFile.nameKey).toEqual('key1')
    expect(raFile.toString()).toEqual(
      input
        .replace('key1 valD', '    key1 valD')
        .replace('key2 valE', '    key2 valE')
        .replace('key3 valF', '    key3 valF'),
    )
  })

  it('throws on an empty stanza', () => {
    expect(() => new RaFile('')).toThrow(/Invalid stanza, was empty/)
    const raFile = new RaFile()
    expect(() => raFile.add('')).toThrow(/Invalid stanza, was empty/)
  })

  it('throws if stanzas have mismatched keys', () =>
    expect(
      () => new RaFile(fs.readFileSync('test/mismatchKeys.ra', 'utf8')),
    ).toThrow(/must have the same key/))

  it('throws if stanzas have duplicate names', () =>
    expect(
      () => new RaFile(fs.readFileSync('test/duplicateName.ra', 'utf8')),
    ).toThrow(/duplicate stanza name/))
})
