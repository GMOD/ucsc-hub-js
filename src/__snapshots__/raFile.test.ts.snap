// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles CRLF newlines 1`] = `
RaFile {
  "data": {
    "valueA": Map {
      "key1" => "valueA",
      "key2" => "valueB",
    },
    "valueC": Map {
      "key1" => "valueC",
      "key2" => "valueD",
    },
  },
  "nameKey": "key1",
}
`;

exports[`handles indented stanzas 1`] = `
RaFile {
  "data": {
    "Foot": Map {
      "key1" => "Foot",
      "key2" => "Ankle",
      "key3" => "Knee",
    },
    "valA": Map {
      "key1" => "valA",
      "key2" => "valB",
      "key3" => "valC",
    },
    "valD": Map {
      "key1" => "valD",
      "key2" => "valE",
      "key3" => "valF",
    },
    "valG": Map {
      "key1" => "valG",
      "key2" => "valH",
      "key3" => "valI",
    },
  },
  "nameKey": "key1",
}
`;

exports[`parses a file with comments 1`] = `
RaFile {
  "data": {
    "valA": Map {
      "key1" => "valA",
      "key2" => "valB",
      "key3" => "valC",
    },
    "valD": Map {
      "key1" => "valD",
      "key2" => "valE",
      "key3" => "valF",
    },
  },
  "nameKey": "key1",
}
`;

exports[`parses a list of stanzas 1`] = `
RaFile {
  "data": {
    "valueA": Map {
      "key1" => "valueA",
      "key2" => "valueB",
    },
    "valueC": Map {
      "key1" => "valueC",
      "key2" => "valueD",
    },
  },
  "nameKey": "key1",
}
`;

exports[`parses a simple three stanza file 1`] = `
RaFile {
  "data": {
    "valA": Map {
      "key1" => "valA",
      "key2" => "valB",
      "key3" => "valC",
    },
    "valD": Map {
      "key1" => "valD",
      "key2" => "valE",
      "key3" => "valF",
    },
    "valG": Map {
      "key1" => "valG",
      "key2" => "valH",
      "key3" => "valI",
    },
  },
  "nameKey": "key1",
}
`;
